kinta-lib/src/main/kotlin/com/dailymotion/kinta/integration/appgallery/AppGalleryIntegration.kt

```diff
@@ -0,0 +1,444 @@
+        }
+        return jsonResult["appids"]!!.jsonArray[0].jsonObject["value"]?.content!!
+    }
+
+    private fun getUploadUrl(
+            clientId: String,
+            token: String,
+            appId: String,
+            fileExtension: String
+    ): UploadData {
+
+        val request = getRequest(clientId, token)
+                .url(HttpUrl.parse("$API_URL/publish/v2/upload-url")!!.newBuilder()
+                        .addQueryParameter("appId", appId)
+                        .addQueryParameter("suffix", fileExtension)
+                        .build()
+                )
+                .build()
+
+        Logger.i("Getting upload url from AppGallery...")
+        val response = OkHttpClient.Builder().build().newCall(request).execute()
+        check(response.isSuccessful) {
+            "Error getting upload url : ${response.code()} ${response.message()}"
+        }
+        val jsonResult = Json.nonstrict.parseJson(response.body()?.string().orEmpty()).jsonObject
+        check(jsonResult["ret"]?.jsonObject?.get("code")?.intOrNull ?: -1 == 0){
+            "Error getting upload url : $jsonResult"
+        }
+        return UploadData(jsonResult["uploadUrl"]!!.content, jsonResult["authCode"]!!.content)
+    }
+
+    private fun uploadFile(
+            clientId: String,
+            token: String,
+            uploadData: UploadData,
+            fileName: String,
+            file: File
+    ): String {
+
+        val okHttp = OkHttpClient.Builder()
+                .callTimeout(Duration.ZERO)
+                .writeTimeout(Duration.ZERO)
+                .readTimeout(Duration.ZERO)
+                .build()
+
+        val mediaType = MediaType.parse("application/vnd.android.package-archive")
+        val request = getRequest(clientId, token)
+                .url(uploadData.uploadUrl)
+                .post(MultipartBody.Builder()
+                        .addFormDataPart("name", fileName)
+                        .addFormDataPart("authCode", uploadData.authCode)
+                        .addFormDataPart("fileCount", 1.toString())
+                        .addFormDataPart("file", fileName, ProgressRequestBody(RequestBody.create(mediaType, file), object : ProgressRequestBody.Listener {
+                            override fun onProgress(progress: Int) {
+                                Logger.d("Upload progress $progress")
+                            }
+                        }))
+                        .build())
+                .build()
+
+        Logger.i("Uploading file to AppGallery...")
+        val response = okHttp.newCall(request).execute()
+        check(response.isSuccessful) {
+            "Error uploading to AppGallery : ${response.body()?.string()}"
+        }
+        val jsonResult = Json.nonstrict.parseJson(response.body()?.string().orEmpty()).jsonObject
+        val filesArray = jsonResult["result"]!!.jsonObject["UploadFileRsp"]!!.jsonObject.getArray("fileInfoList")
```

Comment:
It might be worth using retrofit to handle the Json conversion automagically there ? No strong opinion about this though. Either work for me.